src/Elixir.Khepri.erl

```diff
@@ -0,0 +1,119 @@
+         'put!'/2, 'put!'/3, 'put!'/4, 'put!'/5,
+
+         get/1, get/2, get/3,
+         'get!'/1, 'get!'/2, 'get!'/3,
+
+         delete/1, delete/2, delete/3,
+         'delete!'/1, 'delete!'/2, 'delete!'/3
+        ]).
+
+-compile({no_auto_import, [get/1, get/2, put/2, erase/1]}).
+
+start() ->
+    khepri:start().
+
+'start!'() ->
+    Ret = start(),
+    handle_bang_return_value(Ret).
+
+start(RaSystem) ->
+    khepri:start(RaSystem).
+
+'start!'(RaSystem) ->
+    Ret = start(RaSystem),
+    handle_bang_return_value(Ret).
+
+start(RaSystem, ClusterName, FriendlyName) ->
+    khepri:start(RaSystem, ClusterName, FriendlyName).
+
+'start!'(RaSystem, ClusterName, FriendlyName) ->
+    Ret = start(RaSystem, ClusterName, FriendlyName),
+    handle_bang_return_value(Ret).
+
+put(PathPattern, Data) ->
+    khepri:put(PathPattern, Data).
+
+'put!'(PathPattern, Data) ->
+    Ret = put(PathPattern, Data),
+    handle_bang_return_value(Ret).
+
+put(StoreId, PathPattern, Data) ->
+    khepri:put(StoreId, PathPattern, Data).
+
+'put!'(StoreId, PathPattern, Data) ->
+    Ret = put(StoreId, PathPattern, Data),
+    handle_bang_return_value(Ret).
+
+put(StoreId, PathPattern, Data, ExtraOrOptions) ->
+    khepri:put(StoreId, PathPattern, Data, ExtraOrOptions).
+
+'put!'(StoreId, PathPattern, Data, ExtraOrOptions) ->
+    Ret = put(StoreId, PathPattern, Data, ExtraOrOptions),
+    handle_bang_return_value(Ret).
+
+put(StoreId, PathPattern, Data, Extra, Options) ->
+    khepri:put(StoreId, PathPattern, Data, Extra, Options).
+
+'put!'(StoreId, PathPattern, Data, Extra, Options) ->
+    Ret = put(StoreId, PathPattern, Data, Extra, Options),
+    handle_bang_return_value(Ret).
+
+get(PathPattern) ->
+    khepri:get(PathPattern).
+
+'get!'(PathPattern) ->
+    Ret = get(PathPattern),
+    handle_bang_return_value(Ret).
+
+get(StoreId, PathPattern) ->
+    khepri:get(StoreId, PathPattern).
+
+'get!'(StoreId, PathPattern) ->
+    Ret = get(StoreId, PathPattern),
+    handle_bang_return_value(Ret).
+
+get(StoreId, PathPattern, Options) ->
+    khepri:get(StoreId, PathPattern, Options).
+
+'get!'(StoreId, PathPattern, Options) ->
+    Ret = get(StoreId, PathPattern, Options),
+    handle_bang_return_value(Ret).
+
+delete(PathPattern) ->
+    khepri:delete(PathPattern).
+
+'delete!'(PathPattern) ->
+    Ret = delete(PathPattern),
+    handle_bang_return_value(Ret).
+
+delete(StoreId, PathPattern) ->
+    khepri:delete(StoreId, PathPattern).
+
+'delete!'(StoreId, PathPattern) ->
+    Ret = delete(StoreId, PathPattern),
+    handle_bang_return_value(Ret).
+
+delete(StoreId, PathPattern, Options) ->
+    khepri:delete(StoreId, PathPattern, Options).
+
+'delete!'(StoreId, PathPattern, Options) ->
+    Ret = delete(StoreId, PathPattern, Options),
+    handle_bang_return_value(Ret).
+
+%% TODO: Do we need to "reformat" the error tuple to something more Elixir-y?
+handle_bang_return_value({ok, Result}) -> Result;
+handle_bang_return_value(Error)        -> throw(Error).
```

Comment:
The problem is we introduce a dependency to Elixir's runtime if we call `'Elixir.RuntimeError'`.

Reply:
Ah I think we can just use `error/1`. (`throw/1` is ok too but I think errors are more common to handle in Elixir because that's how `raise/2` and exceptions are implemented.)

```erl
-module(except).

-export([fail/0]).

fail() ->
    error("ah!").
```

```
iex(1)> :except.fail()
** (ErlangError) Erlang error: 'ah!'
    (except 0.1.0) src/except.erl:6: :except.fail/0
```

Which you could rescue as an `ErlangError`.

Yeah good point, let's avoid any runtime dependencies 👍. I think it might also be bad to try to create an exception in Erlang from scratch since some of it seems like implementation details of Elixir (i.e. the hidden `__exception__` key in a map).

Reply:
Thank you, I changed the code to use `erlang:error/1` instead of `erlang:throw/1`.